emission/analysis/modelling/trip_model/greedy_similarity_binning.py

```diff
@@ -0,0 +1,252 @@
+        structure to capture the notion of binning. this was then copied into
+        a Dict when the model needed to be saved. the same technique can be 
+        written to work directly on nested Dicts with no loss in performance. 
+        the data takes the form:
+        {
+            bin_id: {
+                "features": [
+                    [f1, f2, .., fn],
+                    ...
+                ],
+                "labels": [
+                    { label1: value1, ... }
+                ],
+                "predictions": [
+                    { "labels": { label1: value1, ... }, 'p': p_val }
+                ]
+            }
+        }
+        where
+        - bin_id:  str    index of a bin containing similar trips, as a string
+                          (string type for bin_id comes from mongodb object key type requirements)
+        - f_x:     float  feature value (an ordinate such as origin.x)
+        - label_x: str    OpenPATH user label category such as "mode_confirm"
+        - value_x: str    user-provided label for a category
+        - p_val:   float  probability of a prediction, real number in [0, 1]
+
+        :param config: if provided, a manual configuration for testing purposes. these
+                       values should be provided by the config file when running OpenPATH.
+                       see config.py for more details.
+        """
+
+        if config is None:
+            config = eamtc.get_config_value_or_raise('model_parameters.greedy')
+            logging.debug(f'GreedySimilarityBinning loaded model config from file')
+        else:
+            logging.debug(f'GreedySimilarityBinning using model config argument')
+        
+        expected_keys = [
+            'metric',
+            'similarity_threshold_meters', 
+            'apply_cutoff', 
+            'incremental_evaluation'
+        ]
+        for k in expected_keys:
+            if config.get(k) is None:
+                msg = f"greedy trip model config missing expected key {k}"
+                raise KeyError(msg)
+
+        self.metric = eamssmt.SimilarityMetricType.from_str(config['metric']).build()
+        self.sim_thresh = config['similarity_threshold_meters']
+        self.apply_cutoff = config['apply_cutoff']
+        self.is_incremental = config['incremental_evaluation']
+
+        self.bins: Dict[str, Dict] = {}
+        
+
+    def fit(self, trips: List[ecwc.Confirmedtrip]):
+        """train the model by passing data, where each row in the data
+        corresponds to a label at the matching index of the label input
+
+        :param trips: 2D array of features to train from
+        """
+        
+        logging.debug(f'fit called with {len(trips)} trips')
+        unlabeled = list(filter(lambda t: len(t['data']['user_input']) == 0, trips))
+        if len(unlabeled) > 0:
+            msg = f'model.fit cannot be called with unlabeled trips, found {len(unlabeled)}'
+            raise Exception(msg)
```

Comment:
right, i agree, but i also want to prevent more obtuse error messages from happening in case things change in the calling scope. perhaps this is just a compulsion of someone who relies heavily on the type system :-) happy to remove if you think it's excessive or if you're concerned about the runtime hit.